python/cugraph/structure/symmetrize.py

```diff
@@ -75,28 +73,84 @@ def symmetrize_df(df, src_name, dst_name):
     return gdf.groupby(by=[src_name, dst_name], as_index=False).min()
 
 
+def symmetrize_ddf(df, src_name, dst_name, weight_name=None):
+    """
+    Take a COO stored in a distributed DataFrame, and the column names of
+    the source and destination columns and create a new data frame
+    using the same column names that symmetrize the graph so that all
+    edges appear in both directions.
+
+    Note that if other columns exist in the data frame (e.g. edge weights)
+    the other columns will also be replicated.  That is, if (u,v,data)
+    represents the source value (u), destination value (v) and some
+    set of other columns (data) in the input data, then the output
+    data will contain both (u,v,data) and (v,u,data) with matching
+    data.
+
+    If (u,v,data1) and (v,u,data2) exist in the input data where data1
+    != data2 then this code will arbitrarily pick the smaller data
+    element to keep, if this is not desired then the caller should
+    should correct the data prior to calling symmetrize.
+
+    Parameters
+    ----------
+    df : dask_cudf.DataFrame
+        Input data frame containing COO.  Columns should contain source
+        ids, destination ids and any properties associated with the
+        edges.
+    src_name : string
+        Name of the column in the data frame containing the source ids
+    dst_name : string
+        Name of the column in the data frame containing the destination ids
+
+    Examples
+    --------
+    >>> M = cudf.read_csv('datasets/karate.csv', delimiter=' ',
+    >>>                   dtype=['int32', 'int32', 'float32'], header=None)
+    >>> sym_df = cugraph.symmetrize(M, '0', '1')
+    >>> G = cugraph.Graph()
+    >>> G.add_edge_list(sym_df['0]', sym_df['1'], sym_df['2'])
+    """
+    if weight_name:
+        ddf2 = df[[dst_name, src_name, weight_name]]
```

Comment:
Maybe not of immediate concern, but does this support dst_name and src_name as list of columns?

Reply:
If we symmetrize before renumbering then we absolutely need to support a list of columns.  If we symmetrize after renumbering then we could get away with just a single column for each.

Reply:
Are you planning on switching to symmetrizing before renumbering in the MG path? It is an option, but it seems to me that it is faster to symmetrize the COO matrix rather than a bunch of strings. That's what SG and MG design doc show. 

Let me know if this is needed for the first version or if this is something that can be added later as part of the C++ version.

Reply:
I would think symmetrizing after would be more efficient.  The documentation example (line 110 above) shows symmetrizing before calling before calling add_edge_list (as well as the documentation earlier).  That threw me off.

So the documentation in symmetrize.py should be updated (add_edge_list is no longer used).

Just found another problem, however.  Within `graph.py` the calls to symmetrize for SG code will work fine.  However, renumbering is now lazy.  That is, we don't actually renumber until right before executing a graph analytic - so we know whether to renumber based on whether we're executing against a CSR-style or CSC-style graph.

So if multi-gpu symmetrize is going to operate after renumbering, it would need to be part of that lazy evaluation.  The function `compute_renumber_edge_list` is called when the renumbering is triggered.

Reply:
Right, the doc was off. Fixed : 0ff8d63 and 8071c3e

For the lazing eval, do I understand correctly that you suggest adding symmetrize at the end of `compute_renumber_edge_list` ?  

Taking a step back, this is an interesting design discussion because when symmetrize is involved, the CSR vs CSC style becomes irrelevant. So, if this was the motivation for the lazy evaluation, this would only apply to `DiGraph` and could actually be waived for `Graph`. Since only a `Graph` would call symmetrize as part of the ETL automation and the release focus on undirected input (ie. `Graph`) I wonder if we really need to worry about it.

Reply:
The motivation for lazy evaluation was to support a DiGraph.

It is true that with the implementation of symmetrize we could renumber and symmetrize eagerly (instead of lazily) on a Graph.  And our focus is on Graph vice DiGraph this release.  At the time that I shifted to lazy evaluation of renumbering we weren't sure we were going to implement symmetrize in this release.

It seems to me like extra work to remove the lazy evaluation from this release (since it already works) and then add it back in later to support DiGraphs.  It seems like the same amount of work this release, and no work later, to add the symmetrize call in the `compute_renumber_edge_list` so that it would work for either Graph or DiGraph.